File: packages/runar-go/analyzer/sig_analyzer.go
package analyzer

import "fmt"

// analyzeSigHygiene implements §9.
//
// - NO_SIG_CHECK: one warning per reachable path with hasCheckSig=false.
// - CHECKSIG_RESULT_DROPPED: one warning per OP_CHECKSIG or
//   OP_CHECKMULTISIG immediately followed by OP_DROP.
func analyzeSigHygiene(ops []Opcode, paths []ExecutionPath) []Finding {
	var out []Finding
	for _, p := range paths {
		if !p.Reachable || p.HasCheckSig {
			continue
		}
		out = append(out, Finding{
			Severity: SeverityWarning,
			Code:     "NO_SIG_CHECK",
			Message:  "Execution path has no signature verification (OP_CHECKSIG/OP_CHECKMULTISIG)",
			Path:     p.Description,
			HasPath:  true,
		})
	}

	for i := 0; i+1 < len(ops); i++ {
		cur := ops[i]
		nxt := ops[i+1]
		if cur.IsRawSpan || nxt.IsRawSpan {
			continue
		}
		isSigOp := byte(cur.Opcode) == opCheckSig || byte(cur.Opcode) == opCheckMultiSig
		if !isSigOp {
			continue
		}
		if byte(nxt.Opcode) != opDrop {
			continue
		}
		out = append(out, Finding{
			Severity:  SeverityWarning,
			Code:      "CHECKSIG_RESULT_DROPPED",
			Message:   fmt.Sprintf("%s result is dropped by %s — signature check has no effect", cur.Name, nxt.Name),
			Offset:    cur.Offset,
			HasOffset: true,
			Opcode:    cur.Name,
			HasOpcode: true,
		})
	}
	return out
}

File: packages/runar-py/runar/analyzer/sig_analyzer.py
"""Signature-hygiene checks per spec §9."""

from __future__ import annotations

from typing import List

from .types import ExecutionPath, Finding, ParsedOpcode


def analyze_sig_hygiene(
    opcodes: List[ParsedOpcode], paths: List[ExecutionPath]
) -> List[Finding]:
    findings: List[Finding] = []

    # NO_SIG_CHECK: one per reachable path with has_check_sig == False.
    for p in paths:
        if p.reachable and not p.has_check_sig:
            findings.append(
                Finding(
                    severity="warning",
                    code="NO_SIG_CHECK",
                    message=(
                        "Execution path has no signature verification "
                        "(OP_CHECKSIG/OP_CHECKMULTISIG)"
                    ),
                    path=p.description,
                )
            )

    # CHECKSIG_RESULT_DROPPED: OP_CHECKSIG/CHECKMULTISIG immediately
    # followed by OP_DROP.
    for i in range(len(opcodes) - 1):
        op = opcodes[i]
        if op.opcode in (0xac, 0xae):  # OP_CHECKSIG, OP_CHECKMULTISIG
            nxt = opcodes[i + 1]
            if nxt.opcode == 0x75:  # OP_DROP
                findings.append(
                    Finding(
                        severity="warning",
                        code="CHECKSIG_RESULT_DROPPED",
                        message=(
                            f"{op.name} result is dropped by {nxt.name} — "
                            "signature check has no effect"
                        ),
                        offset=op.offset,
                        opcode=op.name,
                    )
                )

    return findings
